src/stirling/source_connectors/socket_tracer/protocols/mongodb/decode.cc

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2018- The Pixie Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <libbson-1.0/bson.h>
+
+#include <rapidjson/document.h>
+#include <string>
+
+#include "src/common/base/base.h"
+#include "src/common/base/utils.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/mongodb/decode.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/mongodb/types.h"
+#include "src/stirling/utils/binary_decoder.h"
+
+namespace px {
+namespace stirling {
+namespace protocols {
+namespace mongodb {
+
+ParseState ProcessOpMsg(BinaryDecoder* decoder, Frame* frame) {
+  PX_ASSIGN_OR(uint32_t flag_bits, decoder->ExtractLEInt<uint32_t>(), return ParseState::kInvalid);
+
+  // Find relevant flag bit information and ensure remaining bits are not set.
+  // Bits 0-15 are required and bits 16-31 are optional.
+  if (flag_bits & checksum_bitmask) {
+    frame->checksum_present = true;
+  }
+  if (flag_bits & more_to_come_bitmask) {
+    frame->more_to_come = true;
+  }
+  if (flag_bits & exhaust_allowed_bitmask) {
+    frame->exhaust_allowed = true;
+  }
```

Comment:
I believe that the frame is processed in a single shot, i.e. we would not re-enter this code with a pointer to the same frame. It seems that these members: checksum present, more to come, and exhaust allowed, can be set unambiguously here, i.e. we do not need to rely on default values. Expressing a bias here against default values, could you please consider updating to the following, e.g.:
```
frame->checksum_present = (flag_bits & checksum_bitmask) == checksum_bitmask;
```